pkg/kgo/source.go

```diff
@@ -16,6 +16,29 @@ import (
 	"github.com/twmb/franz-go/pkg/kmsg"
 )
 
+type recordsPool struct{ p *sync.Pool }
+
+func newRecordsPool() recordsPool {
+	return recordsPool{
+		p: &sync.Pool{New: func() any { return &Record{} }},
+	}
+}
+
+func (p recordsPool) get() *Record {
+	if p.p == nil {
+		return &Record{}
+	}
+	return p.p.Get().(*Record)
+}
+
+func (p recordsPool) put(r *Record) {
+	if p.p == nil {
+		return
+	}
+	*r = Record{} // zero out the record
```

Comment:
don't we lose the allocated memory in the slices `Key`, `Value` and `Headers` by doing this?

Reply:
oh ic, this gets copied from `github.com/twmb/franz-go/pkg/kmsg.Record`. 

so to completely remove allocations we'd probably need to make `kmsg.Record` also use a pool and then here we'd copy by the content of the slices instead of copying the slice header: https://github.com/grafana/franz-go/blob/7dbbd49381d799a65c1b5f9b7ceaba8c4d3994d9/pkg/kgo/source.go#L1792

Reply:
right, my intention in this PR is to lay the groundwork for the reuse process of `*kgo.Record` objects. I'll later address the issue of the allocated buffers in a follow-up PR.